pkg/networkpath/traceroute/common/common_windows.go

```diff
@@ -0,0 +1,184 @@
+	}
+
+	err = windows.SetsockoptInt(s, windows.SOL_SOCKET, windows.SO_RCVTIMEO, 100)
+	if err != nil {
+		windows.Closesocket(s) // nolint: errcheck
+		return nil, fmt.Errorf("failed to set SO_RCVTIMEO: %w", err)
+	}
+	return &Winrawsocket{Socket: s}, nil
+}
+
+// SendRawPacket sends a raw packet to a destination IP and port
+func (w *Winrawsocket) SendRawPacket(destIP net.IP, destPort uint16, payload []byte) error {
+
+	dst := destIP.To4()
+	if dst == nil {
+		return errors.New("unable to parse IP address")
+	}
+	sa := &windows.SockaddrInet4{
+		Port: int(destPort),
+		Addr: [4]byte{dst[0], dst[1], dst[2], dst[3]},
+	}
+	if err := sendTo(w.Socket, payload, 0, sa); err != nil {
+		return fmt.Errorf("failed to send packet: %w", err)
+	}
+	return nil
+}
+
+// ListenPackets listens for matching responses based on the passed in trace information and decoderFunc.
+// If neither decoderFunc receives a matching packet within the timeout, a blank response is returned.
+// Once a matching packet is received by a decoderFunc, it will cause the other decoderFuncs to be
+// canceled, and data from the matching packet will be returned to the caller
+func (w *Winrawsocket) ListenPackets(timeout time.Duration, localIP net.IP, localPort uint16, remoteIP net.IP, remotePort uint16, innerIdentifier uint32, matcherFuncs map[int]MatcherFunc) (net.IP, time.Time, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	ip, finished, err := w.handlePackets(ctx, localIP, localPort, remoteIP, remotePort, innerIdentifier, matcherFuncs)
+	if err != nil {
+		_, canceled := err.(CanceledError)
+		if canceled {
+			log.Trace("timed out waiting for responses")
+			return net.IP{}, time.Time{}, nil
+		}
+		log.Errorf("listener error: %s", err.Error())
+
+		return net.IP{}, time.Time{}, fmt.Errorf("error: %w", err)
+	}
+
+	// return the response
+	return ip, finished, nil
+}
+
+// handlePackets in its current implementation should listen for the first matching
+// packet on the connection and then return. If no packet is received within the
+// timeout or if the listener is canceled, it should return a canceledError
+func (w *Winrawsocket) handlePackets(ctx context.Context, localIP net.IP, localPort uint16, remoteIP net.IP, remotePort uint16, innerIdentifier uint32, matcherFuncs map[int]MatcherFunc) (net.IP, time.Time, error) {
+	buf := make([]byte, 512)
+	for {
+		select {
+		case <-ctx.Done():
+			return net.IP{}, time.Time{}, CanceledError("listener canceled")
+		default:
+		}
+
+		// the receive timeout is set to 100ms in the constructor, to match the
+		// linux side. This is a workaround for the lack of a deadline for sockets.
+		//err := conn.SetReadDeadline(now.Add(time.Millisecond * 100))
+		n, _, err := recvFrom(w.Socket, buf, 0)
+		if err != nil {
+			if err == windows.WSAETIMEDOUT {
+				continue
+			}
+			if err == windows.WSAEMSGSIZE {
+				log.Warnf("Message too large for buffer")
+				continue
```

Comment:
Would this result in the same error on the subsequent `recvFrom` call?

Reply:
No, once we read bytes off the wire for a message, we move on to the next one. If the message is too large for the buffer, we receive a truncated version of the message and this error according to the docs: https://learn.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-recvfrom#remarks